tracing-attributes/src/expand.rs

```diff
@@ -202,8 +208,21 @@ fn gen_block<B: ToTokens>(
     // enter the span and then perform the rest of the body.
     // If `err` is in args, instrument any resulting `Err`s.
     if async_context {
-        let mk_fut = match err_event {
-            Some(err_event) => quote_spanned!(block.span()=>
+        let mk_fut = match (err_event, ret_event) {
+            (Some(err_event), Some(ret_event)) => quote_spanned!(block.span()=>
+                async move {
+                    #ret_event;
+                    match async move { #block }.await {
```

Comment:
> #block appears to be expanded twice here, once inside async move { #block }.await expression and another time indirectly in #ret_event.

Ah 🤦, this obviously needs a fix. It's late already so I won't do this today, but let me leave a quick memo for future reference (not very sure this works).
```rust
let ret = #block;
let ret = async move { #block }.await; // for async context
```
```rust
let ret_event = match args.ret_mode {
        Some(FormatMode::Display) => Some(quote!(tracing::trace!(return = %#ret))),
        Some(FormatMode::Debug) => Some(quote!(tracing::trace!(return = ?#ret))),
        _ => None,
};
```

> Also you only want to capture non-error case, so expanding #ret_event at Ok(x) branch of match with x in it, rather than block, would be more correct.

Make sense.

Reply:
yeah, +1 for @ciuncan's suggestion here; there have been bugs in the past related to accidentally expanding an async block twice, so we definitely shouldn't do that. instead, we should only emit the return event in the `Ok` case of the match here.

Reply:
it would be nice to have tests that catch this and fail if the block is evaluated twice --- currently, none of the tests would fail due to this, because none of the instrumented functions have side effects.

Reply:
This looks OK 👍  Also make sure you also put `ret` into return position.